Refactor: Filter available jobs based on user services

ahmedNaser7 opened the pull request:
Filter available jobs based on craftsman services


Previously, all available jobs were displayed. Now, the jobs are filtered to only show those that match the services provided by the craftsman. This is achieved by:

1.  Fetching the craftsman's services using `getUserServicesUseCase`.
2.  Filtering the `mappedJobs` list based on whether the `serviceType` of a job is present in the craftsman's list of services.
3.  Updating the `availableJobs` in the UI state with the filtered list.


https://github.com/user-attachments/assets/0447d48f-1371-4278-b484-79a1c136f33b

ahmedNaser7 committed:
Refactor: Filter available jobs based on user services

This commit refactors the `CraftsmanHomeViewModel` to filter the list of available jobs.

Previously, all available jobs were displayed. Now, the jobs are filtered to only show those that match the services provided by the craftsman. This is achieved by:

1.  Fetching the craftsman's services using `getUserServicesUseCase`.
2.  Filtering the `mappedJobs` list based on whether the `serviceType` of a job is present in the craftsman's list of services.
3.  Updating the `availableJobs` in the UI state with the filtered list.

ahmedNaser7 committed:
Merge branch 'develop' into refactor/filter-craftsman-request-services

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -239,10 +240,15 @@ class CraftsmanHomeViewModel(
                         imageUrl = screenState.value.craftsmanHomeUiState.userServices[job.serviceId]?.imageUrl.orEmpty()
                     )
                 }.also { mappedJobs ->
+                  val services =  getUserServicesUseCase(
+                        screenState.value.craftsmanHomeUiState.phoneNumber,
+                        isCraftsman = true
+                    ).first().map { it.title }
+                    val filteredJobs = mappedJobs?.filter { service-> services.contains(service.serviceType) }
```
you already have the user services at screenState.value.craftsmanHomeUiState.userServices.keys.toList()
and use the serviceId not the serviceType

Joseph-Sameh-0 replied:
because the serviceType will change based on the translation

ahmedNaser7 replied:
exaclty i still thinking about the translation , but if i depend on state , what if the state is null 
i will block until the state updated

Joseph-Sameh-0 replied:
I'm calling the loadAvailableJobs function in the success of getUserServices

ahmedNaser7 replied:
its wrong to map the keys to list : creenState.value.craftsmanHomeUiState.userServices.keys.toList() ❌
but i dont have problem with screen state

Joseph-Sameh-0 replied:
what is the problem
I converts the iteratable to list

ahmedNaser7 committed:
Merge branch 'develop' into refactor/filter-craftsman-request-services

ahmedNaser7 committed:
Fix: Improve filtering of jobs in CraftsmanHomeViewModel

The filtering of jobs in CraftsmanHomeViewModel has been improved to directly use the user's services titles from the screen state, ensuring accurate job display. Previously, a separate call was made to fetch user services, which could lead to inconsistencies.

Joseph-Sameh-0 committed:
Refactor: Add serviceId to RequestServiceUiState and update filtering logic
